src/stream/src/executor/backfill/arrangement_backfill.rs

```diff
@@ -0,0 +1,530 @@
+    /// Upstream table
+    upstream_table: ReplicatedStateTable<S>,
+
+    /// Upstream with the same schema with the upstream table.
+    upstream: BoxedExecutor,
+
+    /// Internal state table for persisting state of backfill state.
+    state_table: ReplicatedStateTable<S>,
+
+    /// The column indices need to be forwarded to the downstream from the upstream and table scan.
+    output_indices: Vec<usize>,
+
+    progress: CreateMviewProgress,
+
+    actor_id: ActorId,
+
+    info: ExecutorInfo,
+
+    metrics: Arc<StreamingMetrics>,
+}
+
+const CHUNK_SIZE: usize = 1024;
+
+impl<S> ArrangementBackfillExecutor<S>
+where
+    S: StateStore,
+{
+    #[allow(clippy::too_many_arguments)]
+    #[allow(dead_code)]
+    pub fn new(
+        upstream_table: ReplicatedStateTable<S>,
+        upstream: BoxedExecutor,
+        state_table: ReplicatedStateTable<S>,
+        output_indices: Vec<usize>,
+        progress: CreateMviewProgress,
+        schema: Schema,
+        pk_indices: PkIndices,
+        metrics: Arc<StreamingMetrics>,
+    ) -> Self {
+        Self {
+            info: ExecutorInfo {
+                schema,
+                pk_indices,
+                identity: "ArrangementBackfillExecutor".to_owned(),
+            },
+            upstream_table,
+            upstream,
+            state_table,
+            output_indices,
+            actor_id: progress.actor_id(),
+            progress,
+            metrics,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    async fn execute_inner(mut self) {
+        // The primary key columns, in the output columns of the upstream_table scan.
+        // Table scan scans a subset of the columns of the upstream table.
+        let pk_in_output_indices = self
+            .upstream_table
+            .pk_indices()
+            .iter()
+            .map(|&i| self.output_indices.iter().position(|&j| i == j))
+            .collect::<Option<Vec<_>>>()
+            .unwrap();
+        let state_len = pk_in_output_indices.len() + 2; // +1 for backfill_finished, +1 for vnode key.
+        let pk_order = self.upstream_table.pk_serde().get_order_types().to_vec();
+        let upstream_table_id = self.upstream_table.table_id();
+        let mut upstream_table = self.upstream_table;
+
+        let schema = Arc::new(self.upstream.schema().clone());
+
+        let mut upstream = self.upstream.execute();
+
+        // Poll the upstream to get the first barrier.
+        let first_barrier = expect_first_barrier(&mut upstream).await?;
+        self.state_table.init_epoch(first_barrier.epoch);
+
+        let progress_per_vnode = get_progress_per_vnode(&self.state_table, state_len).await?;
+
+        let is_completely_finished = progress_per_vnode
+            .iter()
+            .all(|(_, p)| *p == BackfillProgressPerVnode::Completed);
+        if is_completely_finished {
+            assert!(!first_barrier.is_newly_added(self.actor_id));
```

Comment:
Isn't this also true when `is_completely_finished == false`?

Reply:
If `is_completely_finished == false`, this assertion is not always true.
For instance if all progress is `BackfillProgressPerVnode::NotStarted`. Which could mean it is from new streaming graph, and this assertion will eval to false.

Reply:
Updated https://github.com/risingwavelabs/risingwave/pull/10266/commits/9cc6ad26b19a67e4aa9c7e15246b0e885135f243